dubbo-registry/dubbo-registry-kubernetes/src/main/java/org/apache/dubbo/registry/kubernetes/KubernetesRegistry.java

```diff
@@ -0,0 +1,302 @@
+                                if (action == Action.ADDED || action == Action.DELETED) {
+                                    KUBERNETS_EVENT_EXECUTOR.schedule(() -> {
+                                        final List<URL> urlList = queryUrls(url);
+                                        doNotify(url, notifyListener, urlList);
+                                    }, 5, TimeUnit.SECONDS);
+                                }
+                            }
+
+                            @Override
+                            public void onClose(KubernetesClientException e) {
+                                if (logger.isDebugEnabled()) {
+                                    logger.debug("pod watch closed");
+                                }
+                                if (e != null) {
+                                    logger.error("watcher onClose exception", e);
+                                }
+                            }
+                        }));
+    }
+
+    @Override
+    public void unsubscribe(URL url, NotifyListener listener) {
+        if (isProviderSide(url)) {
+            return;
+        }
+
+        super.unsubscribe(url, listener);
+    }
+
+    @Override
+    public void doUnsubscribe(URL url, NotifyListener notifyListener) {
+        Watch watch = kubernetesWatcherMap.remove(url);
+        watch.close();
+    }
+
+    @Override
+    public boolean isAvailable() {
+        return CollectionUtils.isNotEmpty(getAllRunningService());
+    }
+
+    @Override
+    public void destroy() {
+        super.destroy();
+        Collection<URL> urls = Collections.unmodifiableSet(kubernetesWatcherMap.keySet());
+        urls.forEach(url -> {
+            Watch watch = kubernetesWatcherMap.remove(url);
+            watch.close();
+        });
+        KUBERNETS_EVENT_EXECUTOR.shutdown();
+    }
+
+    private boolean isConsumerSide(URL url) {
+        return url.getProtocol().equals(Constants.CONSUMER_PROTOCOL);
+    }
+
+    private boolean isProviderSide(URL url) {
+        return url.getProtocol().equals(Constants.PROVIDER_PROTOCOL);
+    }
+
+    private void registry(URL url, Pod pod) {
+        JSONObject meta = new JSONObject() {{
+            put(Constants.INTERFACE_KEY, url.getServiceInterface());
+            put(FULL_URL, url.toFullString());
+            putAll(url.getParameters());
+        }};
+        kubernetesClient.pods().inNamespace(pod.getMetadata().getNamespace()).withName(pod.getMetadata().getName())
+                .edit()
+                .editMetadata()
+                .addToLabels(MARK, Constants.DEFAULT_PROTOCOL)
+                .addToAnnotations(serviceKey2UniqId(url.getServiceKey()), meta.toJSONString())
+                .and()
+                .done();
+    }
+
+    private String serviceKey2UniqId(String serviecKey) {
+        return SERVICE_KEY_PREFIX + Integer.toHexString(serviecKey.hashCode());
+    }
+
+    private void unregistry(Pod pod, URL url) {
```

Comment:
registry is a noun. Use register(verb) or unregister(verb) instead.

Reply:
thanks,I'll modify it.

Reply:
What can I do next?

Reply:
> registry is a noun. Use register(verb) or unregister(verb) instead.

What can I do next?